File: src/test/java/com/github/hirsivaja/ip/udp/UdpDatagramTest.java
package com.github.hirsivaja.ip.udp;

import com.github.hirsivaja.ip.IpUtils;
import com.github.hirsivaja.ip.TestUtils;
import com.github.hirsivaja.ip.ipv4.Ipv4Header;
import org.junit.Assert;
import org.junit.Test;

import java.nio.ByteBuffer;

public class UdpDatagramTest {

    @Test
    public void udpTest() {
        byte[] ipv4HeaderBytes = IpUtils.parseHexBinary("45C0003E000000000111CEEB0A000002E0000002");
        Ipv4Header header = Ipv4Header.decode(ByteBuffer.wrap(ipv4HeaderBytes));
        byte[] udpBytes = IpUtils.parseHexBinary("02860286002A60E10001001E0AC8C8660000010000140000000004000004000F0000040100040AC8C866");
        UdpDatagram datagram = UdpDatagram.decode(ByteBuffer.wrap(udpBytes), true, header);

        UdpHeader udpHeader = datagram.udpHeader();
        Assert.assertEquals(646, udpHeader.srcPort());
        Assert.assertEquals(646, udpHeader.dstPort());
        Assert.assertEquals(34, datagram.rawData().length);

        byte[] outBytes = TestUtils.toBytes(datagram);
        Assert.assertArrayEquals(udpBytes, outBytes);
    }

    @Test
    public void instantiationTest() {
        byte[] ipv4HeaderBytes = IpUtils.parseHexBinary("45C0003E000000000111CEEB0A000002E0000002");
        Ipv4Header ipv4Header = Ipv4Header.decode(ByteBuffer.wrap(ipv4HeaderBytes));
        byte[] udpData = IpUtils.parseHexBinary("02860286002A60E10001001E0AC8C8660000010000140000000004000004000F0000040100040AC8C866");
        UdpDatagram datagram1 = new UdpDatagram((short) 7876, (short) 2000, udpData, ipv4Header);
        UdpHeader udpHeader1 = datagram1.udpHeader();

        Assert.assertEquals(7876, udpHeader1.srcPort());
        Assert.assertEquals(2000, udpHeader1.dstPort());
        Assert.assertEquals(50, udpHeader1.len());
        Assert.assertEquals(50, datagram1.length());
        Assert.assertEquals((short) 0xD939, udpHeader1.checksum());
        Assert.assertEquals(42, datagram1.data().length());

        UdpDatagram datagram2 = new UdpDatagram(new UdpHeader((short) 7876, (short) 2000, (short) 50, (short) 0), udpData, ipv4Header);
        UdpHeader udpHeader2 = datagram2.udpHeader();

        Assert.assertEquals(7876, udpHeader2.srcPort());
        Assert.assertEquals(2000, udpHeader2.dstPort());
        Assert.assertEquals(50, udpHeader2.len());
        Assert.assertEquals(50, datagram2.length());
        Assert.assertEquals((short) 0xD939, udpHeader2.checksum());
        Assert.assertEquals(42, datagram2.data().length());
    }
}

File: src/test/kotlin/com/github/hirsivaja/ip/packet/payload/UdpDatagramTest.kt
package com.github.hirsivaja.ip.packet.payload

import com.github.hirsivaja.ip.packet.IpPacket
import com.github.hirsivaja.ip.packet.payload.udp.UdpDatagram
import com.github.hirsivaja.ip.packet.payload.udp.UdpHeader
import kotlin.test.Test

class UdpDatagramTest {

    @Test
    fun ipv4UdpDatagramTest() {
        val packetString = "45C0003E000000000111CEEB0A000002E000000202860286002A60E10001001E0AC8C8660000010000140000000004000004000F0000040100040AC8C866"
        val packet: IpPacket = IpPacket.fromByteString(packetString)
        val encodedString = packet.toByteString()

        val udpDatagram = packet.payload as UdpDatagram
        val udpHeader: UdpHeader = udpDatagram.udpHeader
        assert(646.toUShort() == udpHeader.srcPort)
        assert(646.toUShort() == udpHeader.dstPort)
        assert(34 == udpDatagram.data.length())
        assert(packetString == encodedString)
    }

    @Test
    fun ipv6UdpDatagramTest() {
        val packetString = "600000000042113E200300DE201601FF0000000000000011200300DE20160125FC3683174E86CB7200A1FF5000420FF6303802010104146E35724144316967333134497166696F59425777A21D020455E8832A020100020100300F300D06082B060102010B0500410100"
        val packet: IpPacket = IpPacket.fromByteString(packetString)
        val encodedString = packet.toByteString()

        val udpDatagram = packet.payload as UdpDatagram
        val udpHeader: UdpHeader = udpDatagram.udpHeader
        assert(161.toUShort() == udpHeader.srcPort)
        assert(65360.toUShort() == udpHeader.dstPort)
        assert(58 == udpDatagram.data.length())
        assert(packetString == encodedString)
    }
}
